README.md
# cloudmonkey-interoute (fork from cloudstack-cloudmonkey version 5.3.3)

This is a fork of the [Cloudmonkey command line interface](https://github.com/apache/cloudstack-cloudmonkey), version 5.3.3, modified for use with the [Interoute Virtual Data Centre](https://cloudstore.interoute.com/what_is_vdc).

The modification required is to handle API access to the different VDC regions (Europe, North America (USA), Asia). This is implemented using a new configuration variable `region`.

The default value is 'europe'. Note that the origin version of Cloudmonkey (cloudstack-cloudmonkey) will access the default region without modification.

See the origin repo here: [cloudstack-cloudmonkey](https://github.com/apache/cloudstack-cloudmonkey)

## Modifications

The files `cloudmonkey.py`, `requester.py` and `config.py` have been modified. To access the Asia or USA region, add a line to each required profile setting in the Cloudmonkey `config` file:

    region = asia

or 

    region = usa

(the region name is not case sensitive). You can also change the region interactively while Cloudmonkey is running:

    set region usa


## How to install this modified version

You can make a fresh install of this modified version of Cloudmonkey using the `pip` command:

    sudo pip install git+https://github.com/Interoute/cloudmonkey-interoute.git

Or to upgrade an existing installation:

    sudo pip install --upgrade git+https://github.com/Interoute/cloudmonkey-interoute.git

This version of Cloudmonkey may not work with other cloud computing providers that are also compatible with Cloudmonkey. So be careful if you use Cloudmonkey with several providers.

(The reason is that this version adds a new parameter input 'region=...' to every API call. API servers generally ignore any parameters that are not recognised, however if the server is programmed to do validity checking then Cloudmonkey may not work.)

An alternative is to use [Python's virtualenv](http://www.pythoncentral.io/how-to-install-virtualenv-python) tool so that you can operate multiple versions of Cloudmonkey at the same time.



CHANGES.md
Apache CloudStack CloudMonkey Changelog
---------------------------------------


(2016-11-24) INTEROUTE FORK OF VERSION 5.3.3
============================================
A modification is required for using Cloudmonkey with Interoute Virtual Data Centre. 
This is to handle access to the different VDC regions, by creating a new configuration 
variable 'region'. 
The files cloudmonkey.py, requester.py and config.py have been modified.


Version 5.3.3
=============
This release includes
- Support for shell history
- Sort autocompletion suggestion based on names
- Improve CSV output and error handling


Version 5.3.2
=============
This release includes
- Spinner printing improvements
- Monkey patching SSL requests
- Encode publickey, privatekey and certificates in APIs
- Configurable signature version
- Fix tabular output mode
- Better response checking in the API response
- A new XML display output mode
- A new CSV display output mode

Version 5.3.1
=============
This release includes
- Users can specify domain when using username/password auth per server profile
- Autocompletion of args works when cursor is not at the end of the line
- CLOUDSTACK-7935: keep colons in the request to ACS
- Account parameters are sometimes UUIDs and sometimes string, CloudMonkey
  now automatically autocompletes for both UUID and string account args
- Pass verifysslcert option while user logs in using username/password
- Importing readline no longer outputs escape characters
- CloudMonkey will not output extra empty lines in stdout output
- Filtered result output is uniform across output display formats
- Async blocked API now show a spinning cursor instead of print dots
- When finding missing API args, it does case insensitive search
- New command line arg: -p or --profile (load server profile)
- New command line arg: -b, --block-async (block and poll result on async API calls)
- New command line arg: -n, --noblock-async (do not block on async API calls)

Version 5.3.0
=============
This release includes
- CloudMonkey becomes unicode friendly
- Autocompletion for filters, precache changes
- Autocompletion for config 'set' options and boolean api args
- Current server profile displayed on prompt
- Server profile related bugfixes, blank profile names are not allowed
- Filtering support for default display output
- Filtering by single key only outputs results without key names
- Non-interactive commands from command line are outputted without colors
- Parameter completion uses list api heuristics and related APIs as fallback
- Parameter completion options are cached to speed up rendering of options
- CloudMonkey returns non-zero exit code when run on shell and a error is return
  from managment server, the error message is written to stderr
- Adds new config parameter 'verifysslcert' to enable/disable SSL cert checking
- New command line arg: -d for display (json, table or default)

Make sure you backup your config before you upgrade cloudmonkey from previous releases.
With this release `cloudmonkey` will automatically fix your config file, add missing
configuration parameters and save it as the upgraded versions starts for the first time.

Version 5.2.0
=============
This release includes
 - In the config [server] section is deprecated now
 - For missing keys, cloudmonkey will set default values
 - Network requests, json decoding and shell related bugfixes
 - Based on platform, it will install either pyreadline (Windows) or readline (OSX and Linux)
 - Config options `protocol`, `host`, `port`, `path` are deprecated now
 - Backward compatibilty exists for above options but we use `url` for the mgmt server URL
 - Introduces server profiles so users can use cloudmonkey with different hosts and management server configs
 - A default profile under the section [local] is added with default values
 - Everytime `set` is called, cloudmonkey will write the config and reload config file

Make sure you backup your config before you upgrade cloudmonkey from previous releases.
With this release `cloudmonkey` will automatically fix your config file, add missing
configuration parameters and save it as the upgraded versions starts for the first time.

Version 5.1.0
=============
This release includes
 - support for using username and password instead of / in addition to api key and secret key
 - Usage of signature version 3 for the api signing process. This reduces the chance of API replay attacks
 - cleanup based on reporting from PEP8 and Flake8

If you upgrade from 5.0, then cloudmonkey will ask you to update your config file (~/.cloudmonkey/config)
Under the [user], you can add
username =
password =
Under the [server], you can add
expires = 600

Version 5.0.0
=============
This is the first release of CloudMonkey independent from the Apache CloudStack core orchestration engine. The release
includes a precache of Apache CloudStack 4.2.0 API calls, and should be backward compatible with prior 3.x and 4.x
CloudStack installations (with the obvious caveat that previous versions will have a subset of the latest API commands /
parameters).


cloudmonkey/requester.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

try:
    import base64
    import hashlib
    import hmac
    import itertools
    import json
    import requests
    import ssl
    import sys
    import time
    import urllib
    import urllib2

    from datetime import datetime, timedelta
    from requests_toolbelt import SSLAdapter
    from urllib2 import HTTPError, URLError
except ImportError, e:
    print "Import error in %s : %s" % (__name__, e)
    import sys
    sys.exit()


# Disable HTTPS verification warnings.
from requests.packages import urllib3
urllib3.disable_warnings()


def logger_debug(logger, message):
    if logger is not None:
        logger.debug(message)


def writeError(msg):
    sys.stderr.write(msg)
    sys.stderr.write("\n")
    sys.stderr.flush()


def login(url, username, password, domain="/", verifysslcert=False):
    """
    Login and obtain a session to be used for subsequent API calls
    Wrong username/password leads to HTTP error code 531
    """
    args = {}

    args["command"] = 'login'
    args["username"] = username
    args["password"] = password
    args["domain"] = domain
    args["response"] = "json"

    sessionkey = ''
    session = requests.Session()
    session.mount('https://', SSLAdapter(ssl.PROTOCOL_TLSv1))

    try:
        resp = session.post(url, params=args, verify=verifysslcert)
    except requests.exceptions.ConnectionError, e:
        writeError("Connection refused by server: %s" % e)
        return None, None

    if resp.status_code == 200:
        sessionkey = resp.json()['loginresponse']['sessionkey']
    elif resp.status_code == 405:
        writeError("Method not allowed, unauthorized access on URL: %s" % url)
        session = None
        sessionkey = None
    elif resp.status_code == 531:
        writeError("Error authenticating at %s using username: %s"
                   ", password: %s, domain: %s" % (url, username, password,
                                                   domain))
        session = None
        sessionkey = None
    else:
        resp.raise_for_status()

    return session, sessionkey


def logout(url, session):
    if not session:
        return
    session.get(url, params={'command': 'logout'})


def make_request_with_password(command, args, logger, url, credentials,
                               verifysslcert=False):
    error = None
    args = args.copy()
    username = credentials['username']
    password = credentials['password']
    domain = credentials['domain']

    if not (username and password):
        error = "Username and password cannot be empty"
        result = None
        return result, error

    tries = 0
    retry = True

    while tries < 2 and retry:
        sessionkey = credentials.get('sessionkey')
        session = credentials.get('session')
        tries += 1

        # obtain a valid session if not supplied
        if not (session and sessionkey):
            session, sessionkey = login(url, username, password, domain,
                                        verifysslcert)
            if not (session and sessionkey):
                return None, 'Authentication failed'
            credentials['session'] = session
            credentials['sessionkey'] = sessionkey

        args['sessionkey'] = sessionkey

        # make the api call
        resp = session.get(url, params=args, verify=verifysslcert)
        result = resp.text
        logger_debug(logger, "Response received: %s" % resp.text)

        if resp.status_code == 200:  # success
            retry = False
            break
        if resp.status_code == 401:  # sessionkey is wrong
            credentials['session'] = None
            credentials['sessionkey'] = None
            continue

        if resp.status_code != 200 and resp.status_code != 401:
            error_message = resp.headers.get('X-Description')
            if not error_message:
                error_message = resp.text
            error = "{0}: {1}".format(resp.status_code, error_message)
            result = None
            retry = False

    return result, error


def make_request(command, args, logger, url, credentials, expires, region,
                 verifysslcert=False, signatureversion=3):
    result = None
    error = None

    if not url.startswith('http'):
        error = "Server URL should start with 'http' or 'https', " + \
                "please check and fix the url"
        return None, error

    if not args:
        args = {}

    args = args.copy()
    args["command"] = command
    args["region"] = region
    args["response"] = "json"
    signatureversion = int(signatureversion)
    if signatureversion >= 3:
        args["signatureversion"] = signatureversion
        if not expires:
            expires = 600
        expirationtime = datetime.utcnow() + timedelta(seconds=int(expires))
        args["expires"] = expirationtime.strftime('%Y-%m-%dT%H:%M:%S+0000')

    for key in args.keys():
        value = args[key]
        if isinstance(value, unicode):
            value = value.encode("utf-8")
        args[key] = value
        if not key:
            args.pop(key)

    # try to use the apikey/secretkey method by default
    # followed by trying to check if we're using integration port
    # finally use the username/password method
    if not credentials['apikey'] and not ("8096" in url):
        try:
            return make_request_with_password(command, args, logger, url,
                                              credentials, verifysslcert)
        except (requests.exceptions.ConnectionError, Exception), e:
            return None, e

    def sign_request(params, secret_key):
        request = zip(params.keys(), params.values())
        request.sort(key=lambda x: x[0].lower())
        hash_str = "&".join(
            ["=".join(
                [r[0].lower(),
                 urllib.quote_plus(str(r[1]), safe="*").lower()
                 .replace("+", "%20").replace("%3A", ":")]
            ) for r in request]
        )
        return base64.encodestring(hmac.new(secret_key, hash_str,
                                   hashlib.sha1).digest()).strip()

    args['apiKey'] = credentials['apikey']
    args["signature"] = sign_request(args, credentials['secretkey'])

    session = requests.Session()
    session.mount('https://', SSLAdapter(ssl.PROTOCOL_TLSv1))

    try:
        response = session.get(url, params=args, verify=verifysslcert)
        logger_debug(logger, "Request sent: %s" % response.url)
        result = response.text

        if response.status_code == 200:  # success
            error = None
        elif response.status_code == 401:      # auth issue
            error = "401 Authentication error"
        elif response.status_code != 200 and response.status_code != 401:
            error = "{0}: {1}".format(response.status_code,
                                      response.headers.get('X-Description'))
    except requests.exceptions.ConnectionError, e:
        return None, "Connection refused by server: %s" % e
    except Exception, pokemon:
        error = pokemon.message

    logger_debug(logger, "Response received: %s" % result)
    if error is not None:
        logger_debug(logger, "Error: %s" % (error))
        return result, error

    return result, error

def monkeyrequest(command, args, isasync, asyncblock, logger, url,
                  credentials, timeout, expires, region, verifysslcert=False, signatureversion=3):
    response = None
    error = None
    logger_debug(logger, "======== START Request ========")
    logger_debug(logger, "Requesting command=%s, args=%s" % (command, args))

    response, error = make_request(command, args, logger, url,
                                   credentials, expires, region, verifysslcert, signatureversion)

    logger_debug(logger, "======== END Request ========\n")

    if error is not None and not response:
        return response, error

    def process_json(response):
        try:
            response = json.loads(response, "utf-8")
        except ValueError, e:
            logger_debug(logger, "Error processing json: %s" % e)
            writeError("Error processing json: %s" % e)
            response = None
            error = e
        return response

    response = process_json(response)
    if not response or not isinstance(response, dict):
        return response, error

    m = list(v for v in response.keys() if 'response' in v.lower())
    if not m:
        return response, 'Invalid response received: %s' % response

    isasync = isasync and (asyncblock == "true" or asyncblock == "True")
    responsekey = filter(lambda x: 'response' in x, response.keys())[0]

    if isasync and 'jobid' in response[responsekey]:
        jobid = response[responsekey]['jobid']
        command = "queryAsyncJobResult"
        request = {'jobid': jobid}
        if not timeout:
            timeout = 3600
        timeout = int(timeout)
        cursor = itertools.cycle([u'|', u'/', u'-', u'\\'])
        while timeout > 0:
            interval = 2
            while interval > 0:
                sys.stdout.write(u"%s\r" % cursor.next())
                sys.stdout.flush()
                time.sleep(0.1)
                interval -= 0.1
            timeout = timeout - 2
            logger_debug(logger, "Job %s to timeout in %ds" % (jobid, timeout))
            response, error = make_request(command, request, logger, url,
                                           credentials, expires, region, verifysslcert, signatureversion)
            if error and not response:
                return response, error

            response = process_json(response)
            responsekeys = filter(lambda x: 'response' in x, response.keys())

            if len(responsekeys) < 1:
                continue

            result = response[responsekeys[0]]
            if "errorcode" in result or "errortext" in result:
                return response, error

            jobstatus = result['jobstatus']
            if jobstatus == 2:
                jobresult = result["jobresult"]
                error = "\rAsync job %s failed\nError %s, %s" % (
                        jobid, jobresult["errorcode"], jobresult["errortext"])
                return response, error
            elif jobstatus == 1:
                sys.stdout.write(u"\r \n")
                sys.stdout.flush()
                return response, error
            elif jobstatus == 0:
                pass  # Job in progress
            else:
                logger_debug(logger, "We should not arrive here!")
                sys.stdout.flush()

        error = "Error: Async query timeout occurred for jobid %s" % jobid

    return response, error


cloudmonkey/config.py
# -*- coding: utf-8 -*-
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

__version__ = "5.3.3"
__description__ = "Command Line Interface for Apache CloudStack"
__maintainer__ = "The Apache CloudStack Team"
__maintaineremail__ = "dev@cloudstack.apache.org"
__project__ = "The Apache CloudStack Team"
__projectemail__ = "dev@cloudstack.apache.org"
__projecturl__ = "http://cloudstack.apache.org"

try:
    import os
    import sys

    from ConfigParser import ConfigParser
    from os.path import expanduser
except ImportError, e:
    print "ImportError", e

param_type = ['boolean', 'date', 'float', 'integer', 'short', 'list',
              'long', 'object', 'map', 'string', 'tzdate', 'uuid']

iterable_type = ['set', 'list', 'object']

# cloudmonkey display types
display_types = ["json", "xml", "csv", "table", "default"]

config_dir = expanduser('~/.cloudmonkey')
config_file = expanduser(config_dir + '/config')

# cloudmonkey config fields
mandatory_sections = ['core', 'ui']
default_profile_name = 'local'
config_fields = {'core': {}, 'ui': {}}

# core
config_fields['core']['asyncblock'] = 'true'
config_fields['core']['paramcompletion'] = 'true'
config_fields['core']['cache_file'] = expanduser(config_dir + '/cache')
config_fields['core']['history_file'] = expanduser(config_dir + '/history')
config_fields['core']['log_file'] = expanduser(config_dir + '/log')
config_fields['core']['profile'] = default_profile_name

# ui
config_fields['ui']['color'] = 'true'
config_fields['ui']['prompt'] = '🐵 > '
config_fields['ui']['display'] = 'default'

# default profile
default_profile = {}
default_profile['url'] = 'http://localhost:8080/client/api'
default_profile['timeout'] = '3600'
default_profile['expires'] = '600'
default_profile['username'] = 'admin'
default_profile['password'] = 'password'
default_profile['region'] = 'europe'
default_profile['domain'] = '/'
default_profile['apikey'] = ''
default_profile['secretkey'] = ''
default_profile['verifysslcert'] = 'true'
default_profile['signatureversion'] = '3'


def write_config(get_attr, config_file):
    global config_fields, mandatory_sections
    global default_profile, default_profile_name
    config = ConfigParser()
    if os.path.exists(config_file):
        try:
            with open(config_file, 'r') as cfg:
                config.readfp(cfg)
        except IOError, e:
            print "Error: config_file not found", e

    profile = None
    try:
        profile = get_attr('profile')
    except AttributeError, e:
        pass
    if profile is None or profile == '':
        profile = default_profile_name
    if profile in mandatory_sections:
        print "Server profile name cannot be '%s'" % profile
        sys.exit(1)

    has_profile_changed = False
    profile_in_use = default_profile_name
    try:
        profile_in_use = config.get('core', 'profile')
    except Exception:
        pass
    if profile_in_use != profile:
        has_profile_changed = True

    for section in (mandatory_sections + [profile]):
        if not config.has_section(section):
            try:
                config.add_section(section)
                if section not in mandatory_sections:
                    for key in default_profile.keys():
                        config.set(section, key, default_profile[key])
                else:
                    for key in config_fields[section].keys():
                        config.set(section, key, config_fields[section][key])
            except ValueError, e:
                print "Server profile name cannot be", profile
                sys.exit(1)
        if section in mandatory_sections:
            section_keys = config_fields[section].keys()
        else:
            section_keys = default_profile.keys()
        for key in section_keys:
            try:
                if not (has_profile_changed and section == profile):
                    config.set(section, key, get_attr(key))
            except Exception:
                pass
    with open(config_file, 'w') as cfg:
        config.write(cfg)
    return config


def read_config(get_attr, set_attr, config_file):
    global config_fields, config_dir, mandatory_sections
    global default_profile, default_profile_name
    if not os.path.exists(config_dir):
        os.makedirs(config_dir)

    config_options = reduce(lambda x, y: x + y, map(lambda x:
                            config_fields[x].keys(), config_fields.keys()))
    config_options += default_profile.keys()

    config = ConfigParser()
    if os.path.exists(config_file):
        try:
            with open(config_file, 'r') as cfg:
                config.readfp(cfg)
        except IOError, e:
            print "Error: config_file not found", e
    else:
        config = write_config(get_attr, config_file)
        print "Welcome! Use the `set` command to configure options"
        print "Config file:", config_file
        print "After setting up, run the `sync` command to sync apis\n"

    missing_keys = []
    if config.has_option('core', 'profile'):
        profile = config.get('core', 'profile')
    else:
        global default_profile_name
        profile = default_profile_name

    if profile is None or profile == '' or profile in mandatory_sections:
        print "Server profile cannot be", profile
        sys.exit(1)

    set_attr("profile_names", filter(lambda x: x != "core" and x != "ui",
                                     config.sections()))

    if not config.has_section(profile):
        print ("Selected profile (%s) does not exist," +
               " using default values") % profile
        try:
            config.add_section(profile)
        except ValueError, e:
            print "Server profile name cannot be", profile
            sys.exit(1)
        for key in default_profile.keys():
            config.set(profile, key, default_profile[key])

    for section in (mandatory_sections + [profile]):
        if section in mandatory_sections:
            section_keys = config_fields[section].keys()
        else:
            section_keys = default_profile.keys()
        for key in section_keys:
            try:
                set_attr(key, config.get(section, key))
            except Exception, e:
                if section in mandatory_sections:
                    set_attr(key, config_fields[section][key])
                else:
                    set_attr(key, default_profile[key])
                missing_keys.append("%s = %s" % (key, get_attr(key)))
            # Cosmetic fix for prompt
            if key == 'prompt':
                set_attr(key, get_attr('prompt').strip() + " ")

    if len(missing_keys) > 0:
        print "Missing configuration was set using default values for keys:"
        print "`%s` in %s" % (', '.join(missing_keys), config_file)
        write_config(get_attr, config_file)

    return config_options
